### ISSUE TITLE
EN->CS translation errors: English->Česky, Czechia->Czechia

### ISSUE THREAD

**Issue by martinpopel** [ufal/lindat-translation#44]
There are still many translation errors in CUBBITT, but some are so striking, they are worth of creating an issue (so that I don't forget them).
Two such examples are *English*->*Česky* and *Czechia*->*Czechia* in the (currently default) sentence-level En->Cs model.
* The first one is translated this way only when "English" is the whole sentence. Longer sentences such as "They speak English." seem to be translated correctly ("Mluví anglicky.").
* The second example seems to be translated wrong even in longer contexts, e.g. "I live in Czechia" -> "Žiju v Czechii." (though, in some contexts it is translated as "Čechy" and in some contexts correctly as "Česko").

Luckily, the document-level model (available under Advanced settings) seem to translate both these examples correctly, although I don't have a clear explanation why and I cannot guarantee it is always correct because the error stems from our training data, which is the same for both models (see below my explanation in Czech).

Possible solutions/workarounds:
* Change the default to the doc-level model. The doc-level model is about 5 times slower, though.
* Post-process the translation with some heuristic regexes, e.g.
```perl
if ($src =~ /English/ and $src !~ /Czech/ and $trans =~ /[Čč]esky/) {
  $trans =~ s/Česky/Anglicky/g;
  $trans =~ s/česky/anglicky/g;
}
if ($src =~ /Czechia/ and $trans =~ /Czechi[aeií]/) {
  $trans =~ s/Czechia/Česko/g;
  $trans =~ s/Czechie/Česka/g;
  $trans =~ s/Czechii/Česko/g;
  $trans =~ s/Czechií/Českem/g;
}
```
* An ideal solution would be to (automatically) fix our training data.

Either way, someone should double-check the chosen solution is reasonably reliable and does not have any side effects (so e.g. creating a specialized test set of real-world examples would be a nice first step, e.g. in proper names of companies Czechia->Czechia is the correct translation). We should still watch for similar kind of errors.

**Comment by martinpopel** [comment]
When a user complained about "Czechia->Czechia", this is what I answered:

Děkuji Vám za zpětnou vazbu.
Máte pravdu, že v naprosté většině kontextů se Czechia nepřeloží jako Česko, ale jako Czechia (v Czechii) či Čechy (v Čechách), a že se jedná o chybu. V překladači nemáme žádný ručně sestavený slovník překladů ani pravidla typu if-then-else, je to umělá neuronová síť, která se učí opakovat a zobecňovat to, co viděla v trénovacích datech. V budoucnu možná nějaká pravidla pro opravu častých chyb přidáme, zatím se soustřeďujeme spíš na vylepšení učícího algoritmu a trénovacích dat.

Zde bych mohl odpověď ukončit. Pokud Vás ale zajímá (podobně jako to zaujalo mě, čímž se omlouvám za delší odpověď), proč se překladač naučil překládat Czechia jako Czechia, můžete číst dále.

Sám jsem sice také zastáncem používání krátkého názvu Česko (a Czechia v angličtině) tam, kde stačí zeměpisný název,
ale je pravdou, že názory na toto se různí (viz např. [Spor o užití slova Česko](https://cs.wikipedia.org/wiki/Spor_o_u%C5%BEit%C3%AD_slova_%C4%8Cesko))
a to se odráží i v našich trénovacích datech. Hlavně se to projevuje ve směru z češtiny do angličtiny, kde mnoho Čechů překládá Česko jako Czech Republic, což pak opakuje i náš překladač. Ve směru z angličtiny jsem doufal, že Czechia bude překládána v našich trénovacích datech téměř výhradně jako Česko (tedy i těmi, kdo výraz Czechia nepreferují), ale skutečnost je trochu složitější.

V anglické části paralelních dat je 76 tisíc výskytů Czech Republic, ale jen 555 výskytů Czechia. Z těchto 555 výskytů je distribuce českých překladů přibližně (bez ohledu na pád):
```
345 Česko
 59 Czechia
 58 Česká republika
 34 Čechy
 31 ČR
```

Poznámka na okraj: Ne všechny výskyty "překladu" Czechia jsou špatně. Mohou to být např. vlastní názvy (Medtronic Czechia) nebo věty, které řeší právě otázku názvu naší země ("Již na konci 16. století bylo slovo Czechia natolik běžné, že se dostalo do slovníků"). Stejně tak "ČR" nemusí být špatně - většinou to jsou původně české věty, kde překladatel do angličtiny zvolil Czechia, což je asi přijatelné.

Bohužel z těch 345 "správných" překladů Czechia je většina z textů pochybné kvality, často automaticky generovaných, viz např. 
*"Webkamery British Virgin Islands, Tortola, Britské Panenské ostrovy, Amerika | RACAMERA Global PŘIDAT WEBKAMERU Mobilní verze MAPA O nás Moje webkamery Afrika Amerika Antarktida Asie Austrálie Evropa Oceánie Česko Webkamery a předpověď počasí z British Virgin IslandsAmerikaBritské Panenské ostrovyTortolaBritish Virgin Islands Nanny Cay, Fuel Dock Amerika  Britské Panenské ost...  Tortola  British Virgin Islan...
Charles Vyber vše PŘIDAT WEBKAMERU Mobilní verze MAPA O nás Moje webkamery Afrika Amerika Antarktida Asie Austrálie Evropa Oceánie Česko   copyright © racamera.com"*

Když vynechám věty o webkamerách a bez tečky na konci, zbude jen 166 překladů "Česko" a i mezi nimi je mnoho výčtů názvů států.
Překladač se při trénování učí zohledňovat větný kontext, a tedy se nejspíš naučí, že výčty názvů jsou zvláštní typ vět, který příliš neovlivňuje překlad "normálních" vět.

Stále ale máme v paralelních trénovacích datech jen 59 překladů "Czechia" a přes 100 "Česko" (i v "normálních" větách).
Proč se to tedy překladač nenaučil?

Při trénování používáme metodu tzv. backtranslation, která kvalitu překladu obecně velmi zvyšuje, ovšem zde způsobí chybu.
Backtranslation spočívá v tom, že anglicko-český překladač trénujeme i na původně českých větách, které si automaticky (naším překladačem) přeložíme do angličtiny. V těchto datech je na anglické straně 1281 výskytů "Czechia" a distribuce "překladů" (tedy vlastně originálů) je přibližně:
```
1043 Czechia
 226 Česko
```

To je způsobeno tím, že tyto původně české věty jsme vzali z publicistických textů, kde se v posledních letech velmi diskutovala otázka anglického jednoslovného názvu naší země. Naše trénování kombinuje paralelní trénovací data a tato "syntetická/backtranslation" trénovací data s větším důrazem na ta paralelní, ale v tomto případě nakonec převáží chybný překlad "Czechia".

Pokud by se v anglických textech více používal jednoslovný název Czechia a překladatelé ho konzistentně překládali jako Česko (a naopak Česko do angličtiny jako Czechia a nepřejímali chybné překlady z Lindatu, ale i jiných překladačů) a v českých médiích se ta otázka už tolik neřešila, tak za pár let posbíráme lepší trénovací data a problém se vyřeší sám. Vzhledem k nerealističnosti těchto předpokladů si ale zapisuji Czechia-Česko na seznam vhodných oprav pomocí pravidel, až se k tomu někdy dostanu.
